airbyte-integrations/connectors/source-datascope/source_datascope/source.py

```diff
@@ -0,0 +1,160 @@
+# MIT License
+#
+# Copyright (c) 2020 Airbyte
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+
+from abc import ABC
+from typing import Any, Iterable, List, Mapping, MutableMapping, Optional, Tuple
+from datetime import datetime, timedelta
+import logging
+import backoff
+
+import requests
+from requests.exceptions import ConnectionError
+from requests.structures import CaseInsensitiveDict
+from airbyte_cdk.sources import AbstractSource
+from airbyte_cdk.sources.streams import Stream
+from airbyte_cdk.sources.streams.http import HttpStream
+from airbyte_cdk.sources.streams.http.auth import TokenAuthenticator
+from airbyte_cdk.sources.streams.http.auth import NoAuth
+
+
+
+# Basic full refresh stream
+class DatascopeStream(HttpStream, ABC):
+
+    # TODO: Fill in the url base. Required.
+    url_base = "https://mydatascope.com/api/external/"
+
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        return None
+    @backoff.on_exception(backoff.expo, requests.exceptions.ConnectionError, max_tries=7)
+    def _request(self, url: str, method: str = "GET", data: dict = None) -> List[dict]:
+        response = requests.request(method, url, headers=self._headers, json=data)
+
+        if response.status_code == 200:
+            response_data = response.json()
+            if isinstance(response_data, list):
+                return response_data
+            else:
+                return [response_data]
+        return []
+
+
+class DatascopeIncrementalStream(DatascopeStream, ABC):
+
+    primary_key = "form_answer_id"
```

Comment:
this should not be defined here since the PK is different per stream, it should be defined at the leaf-level stream, WDYT?

Reply:
make sense, going to update that.